README
.. role:: code(strong)
.. role:: file(literal)

======================
README file for Recode
======================

.. contents::
.. sectnum::

Introduction
============

What is Recode?
---------------

Here is version 3.6 for the Recode program and library.  Hereafter,
Recode means the whole package, :code:`recode` means the executable
program.  Glance through this :file:`README` file before starting
configuration.  Make sure you read files :file:`ABOUT-NLS` and
:file:`INSTALL` if you are not familiar with them already.

The Recode library converts files between character sets and usages.
It recognises or produces over 200 different character sets (or about
300 if combined with an :code:`iconv` library) and transliterates files
between almost any pair.  When exact transliteration are not possible,
it gets rid of offending characters or falls back on approximations.
The :code:`recode` program is a handy front-end to the library.

The Recode program and library have been written by François Pinard,
yet it significantly reuses tabular works from Keld Simonsen.  It is an
evolving package, and specifications might change in future releases.

On various Unix systems, Recode is usually compiled from sources,
see the `Installation`_ section below.  On Linux, it often comes
bundled.  Recode had been ported to other popular systems.  See both
`contrib/README`__ and the `Non-Unix ports`_ section below, to find
some more information about these.

__ contrib.html

Reports and collaboration
-------------------------

Send bug reports to mailto:recode-bugs@iro.umontreal.ca' . A bug
report is an adequate description of the problem: your input, what you
expected, what you got, and why this is wrong.  Diffs are welcome, but
they only describe a solution, from which the problem might be uneasy
to infer.  If needed, submit actual data files with your report.  Small
data files are preferred.  Big files may sometimes be necessary, but do
not send them on the mailing list; rather take special arrangement with
the maintainer.

Your feedback will help us to make a better and more portable
package.  Consider documentation errors as bugs, and report them
as such.  If you develop anything pertaining to Recode or have
suggestions, let us know and share your findings by writing at
mailto:recode-forum@iro.umontreal.ca . You may also choose to directly
write at mailto:pinard@iro.umontreal.ca, yet be warned that such
correspondence is often visible for a while through the Recode Web site.

If you feel like receiving releases and pretest announcements for the
Recode package, send a message to mailto:majordomo@iro.umontreal.ca
having, in its body, a line saying::

    subscribe recode-announce

If you rather want to participate actively in discussions, pretesting
and development for Recode, do just as above, but this time, use::

    subscribe recode-forum

Visit http://recode.progiciels-bpi.ca/ for releases or pretests, and related
files.  In particular, button ``Browse`` gives access to a weekly mirror
of the current unpackaged work files, while button ``Folders`` gives
access to saved or pending correspondence.

Please *do not* widely redistribute releases having a letter after the
version numbers, as these are meant for pretesting only, and might not
be stable enough for other usages.

Development plan
----------------

My plan has long been to end the 3.x series of this package, rather
aiming 4.0 as a major internal rewrite.  As there is still a long
way before 4.0 gets ready, and *especially* because some of my good
collaborators insisted that I do so, there will be a Recode 3.7.  That
release is meant to provide a selection of user-contributed patches.

For prototyping what Recode will become and experimenting new concepts
more easily, I created a subsidiary and standalone project named
Recodec, meant to receive the best part of my development efforts in
this particular area.  Once I'll be happy with the prototype, the plan
is to rewrite it from Python to C, somehow.  Visit the Web pages for
this `Recodec project`__ for more information and details.  For now at
least, new features go to Recodec only.

__ http://recodec.progiciels-bpi.ca

Notes for version 3.7-beta2
---------------------------

Here are a few notes related to the beta2 pre-test release for the incoming
Recode 3.7.  I publish it to ease later exchanges of patches with testers.

+ The name has been changed from Free recode to Recode -- as "Free" was
  a four letter word to some people :-). :code:`recode` (no capital) still
  names the executable program specifically, or the distribution archive
  itself.

+ Recode does not itself include :code:`libiconv` anymore.  However,
  it uses an external :code:`iconv` library if one is available at
  installation time, like :code:`libiconv` or the one provided within GNU
  :code:`libc`.  The ``-x:`` option to the program, or a new flag to the
  library :code:`recode_new_outer` function, inhibits the initialisation
  and usage of :code:`iconv`.

+ The bug about loosing a few characters, here and there, when recoding
  big files in :code:`iconv` context, seems to have been corrected.  A
  patch for this problem has been floating around for years, but it was
  not solving all cases.

+ Recode installation now uses Python.  In particular, it creates
  file :file:`build/src/iconvdecl.h` from local ``iconv -l`` output.
  Recode testing through ``make check`` also needs what people
  :code:`python-devel`, providing C header files for Python and
  :code:`distutils`.  The :file:`Makemore` file has been merged within
  regular Makefiles and is not distributed separately anymore.

+ It is likely that new bugs have been introduced through the above
  changes.  In particular, not everything is cosy on the side of release
  engineering.  A few files are either spuriously remade, or remade late.
  I'm a bit surprised by the difficulty to get this right.

+ ``make check`` accepts a ``LIMIT=`` option, for limiting tests to one or
  a few cases.  See :file:`tests/Makefile` for more information.

+ PO files have been updated from the Translation Project.

Notes for version 3.7-beta1
---------------------------

The beta 1 pre-test release for the incoming Recode 3.7 has been made
available for those needing it right away.  While it solves some serious
bugs and portability problems, others are meant to be addressed only in
later pre-tests.  In particular, none of charset or surface issues, user
requests, and various suggestions appear in this pre-test, and will not
either in later pretests, until all real show-stoppers are solved first.
So this is in no way a candidate for a Recode 3.7 release.

The test suite is worth more comments:

+ The suite is very partial, and may not be thought as a validation
  suite.  Before it could be used to ascertain confidence, it would need
  much more tests than it has already.

+ Testing is notably more speedy than it used to be.  For example, the
  previous :code:`bigauto` test, which was not run by default because it
  ran for too long, is now executed within the standard test suite, once
  in non-strict mode, and a second time in strict mode.

+ It does not use Autotest anymore, but rather a home grown test driver
  much inspired from the Codespeak project.  The link between the test and
  the Recode library is established through a Pyrex interface, so you need
  to have :code:`python` and :code:`python-devel` installed first.

+ Beware that the Pyrex interface to the Recode library is only meant
  for testing. for now at least.  While you may play with it, it would not
  be wise relying on it, as the specifications might change at any time.

Installation
============

Prerequisites
-------------

Simple installation of Recode requires the usual tools and facilities as
those needed for most GNU packages.  If not already bundled with your
system, you also need to pre-install Python, version 2.2 or better.  You
may get it from:

  http://www.python.org

It is also convenient to have some :code:`iconv` library already present
on your system, this much extends Recode capabilities, especially in
the area of Asiatic character sets.  GNU :code:`libc`, as found on
Linux systems and a few others, already has such an :code:`iconv`
library.  Otherwise, you might consider pre-installing the portable
:code:`libiconv`, written by Bruno Haible.  You may get it from:

  http://www.gnu.org/software/libiconv/

Getting a release
-----------------

The canonical distribution point for this version is:

  http://recode.progiciels-bpi.ca/archives/recode.tar.gz

GNU mirrors usually hold a copy of non-pretest releases, the canonical
distribution point for the last such release is:

  ftp://ftp.gnu.org/pub/gnu/recode/recode-3.6.tar.gz

Some older distributions *might* be available in this directory:

  http://recode.progiciels-bpi.ca/archives/

Getting the sandbox
-------------------

Visit http://github.com/pinard/Recode/tree/dev-3.7 and use the
*Download* button to get a packaged copy of development sources.  If you
happen to be a Git lover, you may rather use::

  git clone git://github.com/pinard/Recode.git
  
and then, checkout branch ``dev-3.7``.  File timestamps after checktou
may trigger Make difficulties.  As a way to avoid, from the top level of
the distribution, execute ``sh after-patch.sh``.  If you miss either
:code:`sh` or GNU :code:`touch`, try ``python after-patch.py`` instead.

Configure options
-----------------

Once you have an unpacked distribution, see files:

  ===================  =======================================================
  File name            Description
  ===================  =======================================================
  :file:`ABOUT-NLS`    how to customise this program to your language
  :file:`COPYING`      copying conditions for the program
  :file:`COPYING.LIB`  copying conditions for the library
  :file:`INSTALL`      compilation and installation instructions
  :file:`NEWS`         major changes in the current release
  :file:`THANKS`       partial list of contributors
  ===================  =======================================================

Besides those configure options documented in files :file:`INSTALL`
and :file:`ABOUT-NLS`, a few extra options may be accepted after
``./configure``:

+ Options ``--disable-shared`` or ``--disable-static``

  to inhibit the building of shared libraries or static libraries; the
  default is to always build static libraries, and to attempt building
  shared libraries if there is some known recipe for this.

+ Option ``--with-gnu-ld``

  to force the assumption that the C compiler uses GNU ld.

+ Option ``--with-dmalloc``

  to trigger a debugging feature for looking at memory management
  problems, it pre-requires Gray Watson's package, which is available as
  ftp://ftp.letters.com/src/dmalloc/dmalloc.tar.gz .

Maintenance tools
-----------------

For simple modifications to Recode, you should not need special tools
beyond those usual for installing GNU packages.  However, if you modify
any :file:`.l` source file, Python and Flex are both needed for remaking
:file:`merged.c`.

For more comprehensive modifications, you might need more tools.  If not
done already, make sure you have a copy of the packages listed in the
following table.  You may also choose to establish a link in your build
:file:`doc/` directory, as explained within :file:`doc/Makemore`.

  ================   ==========   ==========   =============
  Package name       Current      Minimum      Install after
  ================   ==========   ==========   =============
  :code:`autoconf`   2.61         2.12         :code:`m4`
  :code:`automake`   1.10         1.9          :code:`Perl`
  :code:`Flex`       2.5.33       2.5.4a
  :code:`gettext`    0.16         0.16
  :code:`Help2man`   1.36         1.020        :code:`Perl`
  :code:`libtool`    1.5.24       1.3.4
  :code:`m4`         1.4.10       1.4n
  :code:`Make`       3.81
  :code:`Perl`       5.8.8        5.005.03
  :code:`Python`     2.5.1        2.2
  :code:`tar`        1.17         1.12
  :code:`wget`       1.10.2
  ================   ==========   ==========   =============

The *current* version numbers just happen to be those used for
development, it is often likely that older versions would work just as
well.  The *minimum* version numbers were once acceptable, they might
not be anymore, this has not been verified; any updating information is
welcome!

Installation hints
------------------

Here are a few hints which might help installing Recode on some systems.
Many may be applied by temporary presetting environment variables while
calling ``./configure``.  File :file:`INSTALL` explains this.

+ Compilation time

  Some C compilers, like Apollo's, have a hard time compiling
  :file:`merged.c`.  If this is your case, avoid compiler
  optimisation.  From within the Bourne shell, you may use::

    CFLAGS= ./configure

  But if you want to give a real hard time to your C optimiser on
  :file:`merged.c`, to get code that runs only a bit faster, merely
  try::

      CPPFLAGS=-DINLINE_HARDER ./configure

+ Smallish systems

  For 80286 based systems (do some still exist?!), it has been
  reported that some compilers generate wrong code while optimising
  for *small* models.  So, from within the Bourne shell, do::

      CFLAGS=-Ml LDFLAGS=-Ml ./configure

  to force large memory model.  For 80286 Xenix compiler, the last time
  it was tried a while ago, one ought to use::

      CFLAGS='-Ml -F2000' LDFLAGS=-Ml ./configure

  Other systems have poor :code:`pipe`/:code:`popen` support or thrash
  heavily when processes fork.  In this case, just before doing
  ``make``, edit :file:`config.h` and ensure :code:`HAVE_PIPE` is
  *not* defined.

External pointers
=================

Documentation
-------------

+ IETF references

  + Character Mnemonics & Character Sets

    + ftp://nic.ddn.mil/rfc/rfc1345.txt

    Keld Simonsen <keld@dkuug.dk>, 1992-06.

  + UTF-7 - A Mail-Safe Transformation Format of Unicode

    + ftp://nic.ddn.mil/rfc/rfc1642.txt

    David Goldsmith <david_goldsmith@taligent.com> and Mark Davis
    <ark_davis@taligent.com>, 1994-07.

  +  UTF-8, a transformation format of Unicode and ISO 10646

     + ftp://nic.ddn.mil/rfc/rfc2044.txt

     François Yergeau <yergeau@alis.com>, 1997-10.

+ Various references

  + Unicode charset mappings

    + ftp://ftp.unicode.org:/Public/MAPPINGS/

    The Unicode consortium makes available plenty of charset mappings
    for converting "legacy" charsets to Unicode.

  +  Normalisation et internationalisation: Inventaire et prospectives des
     normes clefs pour le traitement informatique du français.  (392p.)

     This is a report, written in French, discussing charset issues and many
     other topics as well.  Laurent Bourbeau <bourbeau@progiciels-bpi.ca>
     and François Pinard <pinard@iro.umontreal.ca>, 1995-10.

     + ftp://ftp.iro.umontreal.ca/pub/contrib/pinard/accents/oqil-tome1.ps.gz
     + http://www.ceveil.qc.ca/Normes

+ Recode specific

  + ETL presentation

    In 1999, the organisers of the `m17n99 conference`__ in Tsukuba,
    Japan, were kind enough to invite me.  This has been for me a
    fabulous trip and experience, and I met many extraordinary people in
    there.  At the conference, I presented the Translation Project, and
    Recode.  The Recode `presentation slides`__ are available.

__ http://www.m17n.org/conference/m17n99_all_but_registration/welcome.en.html
__ m17n99.html

Programs
--------

+ :code:`libiconv`

  This comprehensive charset converter library revolves around Unicode,
  and support Asian encodings among many others.  Even Recode uses it!

  + http://www.gnu.org/software/libiconv/

  Bruno Haible <haible@ilog.fr>

+ :code:`tcs`

  Here is the main recoding tool from the Plan9 project.

  + ftp://research.att.com/dist/tcs.shar.Z

+ :code:`yuedit`

  This GUI editor handles many encodings, among which UTF-8.  It also
  installs uniconv, a recoding program, and uniprint, a printing tool.

  + ftp://sunsite.unc.edu/pub/Linux/apps/editors/X/yudit-1.2.tar.gz

  Gaspar Sinai <gsinai@iname.com>, 1999-01.

+ :code:`ucs-fonts`

   These 6x13 fonts, covering Unicode characters besides the Asian sets,
   merely replace the Linux fixed 6x13 font.  Works nicely with yudit.

   + http://www.cl.cam.ac.uk/~mgk25/download/ucs-fonts.tar.gz

   Markus Kuhn <Markus.Kuhn@cl.cam.ac.uk>, 1998-11.

+ :code:`MtRecode`

  This charset converter is oriented towards SGML text manipulation.  It
  may be freely downloaded for non-commercial, non-military use from:

  + http://www.lpl.univ-aix.fr/projects/multext/MtRecode/

  Pointer given by Jean Véronis <veronis@univ-aix.fr>, 1996-06.

+ :code:`sp`

  This quite nice SGML structure analyser contains internal C++ modules
  for handling many charsets.

  + ftp://ftp.jclark.com/pub/sp/sp-1.3.tar.gz

  James Clark <jjc@jclark.com>

+ :code:`b2c`

  This program is able to generate interpreted
  character dumps, but properly embedded within complete C header files.

  + http://research.de.uu.net:8080/~gnu/b2c/b2c-2.1.tar.gz

  Jörg Heitkötter <Joerg.Heitkoetter@de.uu.net>, 1997-11.

+ :code:`PyRecode`

  This wrapper provides Recode functionality to Python programs.

  + http://www.suxers.de/PyRecode.tgz

  Andreas Jung <ajung@server.python.net>

  Also see:

  + http://www.vex.net/parnassus/apyllo.py?find=recode
  + http://www.suxers.de/python/pyrecode.htm

Non-Unix ports
--------------

Please mailto:recode-bugs@iro.umontreal.ca if you are aware of various
ports to non-Unix systems not listed here, or for corrections.  Please
provide the goal system, a complete and stable URL, the maintainer name

NEWS
==================================
Recode NEWS - User visible changes
==================================

.. contents::
.. sectnum::

:Copyright: © 1993-1999, 2000, 2001, 2008 Free Software Foundation, Inc.

Version 3.7-beta2
=================

:Author: François Pinard, 2008-03.

+ Recode does no include libiconv anymore, but uses an external iconv
  library if one was available at installation time.
+ Many internal changes, for correcting reported bugs.

Version 3.6
===========

:Author: François Pinard, Bruno Haible, 2001-01.

General changes
---------------

+ The recode manual is now indexed, by charset, by concept, etc.
+ Program messages are also available in Greek, Gallicean and Italian.
+ Bruno Haible's nice portable iconv library has been integrated.
+ RFC 1345 tables and French character names have been updated.
+ The Texinfo charset has been refreshed, and made reversible.

New charsets
------------

(most from libiconv)

+ Japanese

  + EUC-JP (csEUCPkdFmtJapanese, EUC_JP,
    Extended_UNIX_Code_Packed_Format_for_Japanese);
  + ISO-2022-JP (csISO2022JP); ISO-2022-JP-1; ISO-2022-JP-2 (csISO2022JP2);
  + JIS_C6220-1969-ro (csISO14JISC6220ro, ISO646-JP, iso-ir-14, jp);
  + JIS_X0201 (csHalfWidthKatakana, JIS0201, JISX0201-1976, JISX0201.1976-0,
    X0201);
  + JIS_X0208 (csISO87JISX0208, ISO-IR-87, JIS0208, JIS_X0208.1983-0,
    JIS_X0208.1983-1, JIS_X0208-1990-0, JIS_X0208.1983-1, X0208);
  + JIS_X0212 (csISO159JISX02121990, ISO-IR-159, JIS0212, JIS_X0212.1990-0,
    JIS_X0212-1990, X0212);
  + SJIS (csShiftJIS, MS_KANJI, SHIFT-JIS).

+ Chinese

  + BIG5 (BIG-5, BIG-FIVE, BIGFIVE, CN-BIG5 csBig5); BIG5HKSCS;
  + EUC-CN (CN-GB, csGB2312, EUC_CN, GB2312); EUC-TW (csEUCTW, EUC_TW);
  + GB18030; HZ (HZ-GB-2312); ISO-2022-CN (csISO2022CN); ISO-2022-CN-EXT;
  + GB_1988-80 (cn, csISO57GB1988, ISO646-CN, iso-ir-57);
  + GB_2312-80 (CHINESE, csISO58GB231280, GB2312.1980-0, ISO-IR-58);
  + ISO-IR-165 (CN-GB-ISOIR165).

+ Korean

  + JOHAB (CP1361); EUC-KR (csEUCKR, EUC_KR); GBK (CP936);
  + ISO-2022-KR (csISO2022KR);
  + KSC_5601 (CP949, csKSC56011987, ISO-IR-149, KOREAN, KSC5601.1987-0,
    KS_C_5601-1987, KS_C_5601-1989, KSX1001:1992).

+ Vietnamese (independently of libiconv)

  + TCVN; VIQR; VISCII; VNI; VPS.

+ Other languages

  + ARMSCII-8; Georgian-Academy; Georgian-PS; WINDOWS-874 (CP874);
  + MuleLao-1; CP1133 (IBM-CP1133); CP1258 (WINDOWS-1258);
  + TIS-620 (ISO-IR-166, TIS620, TIS620.2529-1, TIS620-0, TIS620.2533-0,
    TIS620.2533-1).

+ Apple specifics

  + MacArabic; MacCentralEurope; MacCroatian; MacCyrillic; MacGreek;
  + MacHebrew; MacIceland; MacRomania; MacThai; MacTurkish; MacUkraine

+ Unicode

  + JAVA; UCS-2-INTERNAL; UCS-2LE (UnicodeLITTLE); UCS-2-SWAPPED; UCS-4BE;
  + UCS-4-INTERNAL; UCS-4LE; UCS-4-SWAPPED; UTF-16BE; UTF-16LE.

+ Others

  + CP932; CP949 (UHC); CP950; CP866 (866, csIBM866, IBM866).
  + ISO-8859-16 (ISO-IR-226, ISO_8859-16:2000).

+ Recode internal

  + :libiconv: (:)   [so option -x: avoids going through libiconv]

New aliases
-----------

(from libiconv) [list to be revised]

+ csASCII (for ANSI_X3.4-1968); csHPRoman8 (for hp-roman8);
+ csISOLatin1 (for ISO-8859-1); csISOLatin2 (for ISO-8859-2);
+ csISOLatin3 (for ISO-8859-3); csISOLatin4 (for ISO-8859-4);
+ csISOLatin5 (for ISO-8859-9);
+ csISOLatin6 and ISO_8859-10:1992 (for ISO-8859-10);
+ csISOLatinArabic (for ISO-8859-6); csISOLatinCyrillic (for ISO-8859-5);
+ csISOLatinGreek (for ISO-8859-7); csISOLatinHebrew (for ISO-8859-8);
+ csKOI8R (for KOI8-R); csPC850Multilingual (for IBM850);
+ csUCS4 (for ISO-10646-UCS-4);
+ csUnicode, csUnicode11, UCS-2BE, UnicodeBIG (for ISO-10646-UCS-2);
+ csUnicode11UTF7 (for UNICODE-1-1-UTF-7);
+ csVISCII and VISCII1.1-1 (for VISCII);
+ ISO-IR-179 (for ISO-8859-13); csMacintosh and MacRoman (for macintosh);
+ TCVN5712-1, TCVN5712-1:1993 and TCVN-5712 (for TCVN).

New surfaces
------------

+ tree (experimental).

Version 3.5
===========

:Author: François Pinard, 1999-05.

Incompatible changes
--------------------

+ A double dot ``..`` should now be used instead of a colon ``:``.
+ Option --force (-f) is needed to pursue recoding despite errors.
+ There is no more quoting for special characters within charsets names.
+ Auto check (``-a``) and popen (``-o``) options have been withdrawn.
+ Some charsets and aliases were deleted, see `Charsets & aliases`_ below.

Extended features
-----------------

+ Program messages are available in localised form for many languages.
+ Long character names are available in French, if LANGUAGE is set to ``fr``.
+ A new request syntax allows for recode chaining, and for surfaces.
+ Option --header-file (-h) accepts a language parameter, and Perl is new.
+ Full charset listings now show the UCS-2 value for characters.
+ Option --known=PAIRS (-k) also accepts octal and hexadecimal numbers.
+ Option --list (-l) better sorts charsets and aliases, also fully written.
+ Charset ``RFC1345`` implements mnemonic+ascii+38, and is now reversible.
+ HTML is not limited anymore to Latin-1, HTML 4.0 entities are supported.

New features
------------

+ Euro support.
+ Updated RFC 1345 set of tables, from Keld Simonsen.
+ Some African charsets and transliterated forms.
+ Conversions for ISO 10646 and Unicode.
+ Combining or explosion of UCS-2 diacriticized characters and ligatures.
+ Implementation of surfaces, see `Surfaces & aliases`_ below.
+ Mixed mode for recoding only comments and strings in C sources or PO files.
+ A stand-alone recoding library gets installed, often as a shared library.
+ Option --find-subsets (-T) lists charsets which are subsets of another.
+ The library may generate testing data, and study character frequencies.

Charsets & aliases
------------------

+ New ISO 10646 and Unicode charsets

  + combined-UCS-2: pseudo-charset.
  + count-characters: pseudo-charset.
  + dump-with-names: pseudo-charset.
  + ISO-10646-UCS-2 (UNICODE-1-1, BMP, rune, u2).
  + ISO-10646-UCS-4 (10646, ISO-10646, UCS-4, u4).

ebcdic.c
/* Conversion of files between different charsets and usages.
   Copyright (C) 1990, 1993 Free Software Foundation, Inc.
   Francois Pinard <pinard@iro.umontreal.ca>, 1988.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#define ASCII "ASCII"
/* Previously: #define ASCII "ASCII-BS" */

#include "recode.h"

static unsigned char const ascii_to_ebcdic[256] =
  {
      0,   1,   2,   3,  55,  45,  46,  47,     /*   0 -   7 */
     22,   5,  37,  11,  12,  13,  14,  15,     /*   8 -  15 */
     16,  17,  18,  19,  60,  61,  50,  38,     /*  16 -  23 */
     24,  25,  63,  39,  28,  29,  30,  31,     /*  24 -  31 */
     64,  79, 127, 123,  91, 108,  80, 125,     /*  32 -  39 */
     77,  93,  92,  78, 107,  96,  75,  97,     /*  40 -  47 */
    240, 241, 242, 243, 244, 245, 246, 247,     /*  48 -  55 */
    248, 249, 122,  94,  76, 126, 110, 111,     /*  56 -  63 */
    124, 193, 194, 195, 196, 197, 198, 199,     /*  64 -  71 */
    200, 201, 209, 210, 211, 212, 213, 214,     /*  72 -  79 */
    215, 216, 217, 226, 227, 228, 229, 230,     /*  80 -  87 */
    231, 232, 233,  74, 224,  90,  95, 109,     /*  88 -  95 */
    121, 129, 130, 131, 132, 133, 134, 135,     /*  96 - 103 */
    136, 137, 145, 146, 147, 148, 149, 150,     /* 104 - 111 */
    151, 152, 153, 162, 163, 164, 165, 166,     /* 112 - 119 */
    167, 168, 169, 192, 106, 208, 161,   7,     /* 120 - 127 */
     32,  33,  34,  35,  36,  21,   6,  23,     /* 128 - 135 */
     40,  41,  42,  43,  44,   9,  10,  27,     /* 136 - 143 */
     48,  49,  26,  51,  52,  53,  54,   8,     /* 144 - 151 */
     56,  57,  58,  59,   4,  20,  62, 225,     /* 152 - 159 */
     65,  66,  67,  68,  69,  70,  71,  72,     /* 160 - 167 */
     73,  81,  82,  83,  84,  85,  86,  87,     /* 168 - 175 */
     88,  89,  98,  99, 100, 101, 102, 103,     /* 176 - 183 */
    104, 105, 112, 113, 114, 115, 116, 117,     /* 184 - 191 */
    118, 119, 120, 128, 138, 139, 140, 141,     /* 192 - 199 */
    142, 143, 144, 154, 155, 156, 157, 158,     /* 200 - 207 */
    159, 160, 170, 171, 172, 173, 174, 175,     /* 208 - 215 */
    176, 177, 178, 179, 180, 181, 182, 183,     /* 216 - 223 */
    184, 185, 186, 187, 188, 189, 190, 191,     /* 224 - 231 */
    202, 203, 204, 205, 206, 207, 218, 219,     /* 232 - 239 */
    220, 221, 222, 223, 234, 235, 236, 237,     /* 240 - 247 */
    238, 239, 250, 251, 252, 253, 254, 255,     /* 248 - 255 */
  };

static unsigned char const ascii_to_ebcdic_ccc[256] =
  {
      0,   1,   2,   3,  55,  45,  46,  47,     /*   0 -   7 */
     22,   5,  37,  11,  12,  13,  14,  15,     /*   8 -  15 */
     16,  17,  18,  19,  60,  61,  50,  38,     /*  16 -  23 */
     24,  25,  63,  39,  28,  29,  30,  31,     /*  24 -  31 */
     64,  90, 127, 123,  91, 108,  80, 125,     /*  32 -  39 */
     77,  93,  92,  78, 107,  96,  75,  97,     /*  40 -  47 */
    240, 241, 242, 243, 244, 245, 246, 247,     /*  48 -  55 */
    248, 249, 122,  94,  76, 126, 110, 111,     /*  56 -  63 */
    124, 193, 194, 195, 196, 197, 198, 199,     /*  64 -  71 */
    200, 201, 209, 210, 211, 212, 213, 214,     /*  72 -  79 */
    215, 216, 217, 226, 227, 228, 229, 230,     /*  80 -  87 */
    231, 232, 233,  74, 224,  79,  95, 109,     /*  88 -  95 */
    121, 129, 130, 131, 132, 133, 134, 135,     /*  96 - 103 */
    136, 137, 145, 146, 147, 148, 149, 150,     /* 104 - 111 */
    151, 152, 153, 162, 163, 164, 165, 166,     /* 112 - 119 */
    167, 168, 169, 192, 106, 208, 161,   7,     /* 120 - 127 */
      4,   6,   8,   9,  10,  20,  21,  23,     /* 128 - 135 */
     26,  27, 138, 139, 140, 141, 142, 143,     /* 136 - 143 */
     32,  33,  34,  35,  36,  40,  41,  42,     /* 144 - 151 */
     43,  44, 154, 155, 156, 157, 158, 159,     /* 152 - 159 */
     48,  49,  51,  52,  53,  54,  56,  57,     /* 160 - 167 */
     58,  59, 170, 171, 172, 173, 174, 175,     /* 168 - 175 */
    176, 177, 178, 179, 180, 181, 182, 183,     /* 176 - 183 */
    184, 185, 186, 187, 188, 189, 190, 191,     /* 184 - 191 */
    128,  65,  66,  67,  68,  69,  70,  71,     /* 192 - 199 */
     72,  73, 202, 203, 204, 205, 206, 207,     /* 200 - 207 */
    144,  81,  82,  83,  84,  85,  86,  87,     /* 208 - 215 */
     88,  89, 218, 219, 220, 221, 222, 223,     /* 216 - 223 */
    160, 225,  98,  99, 100, 101, 102, 103,     /* 224 - 231 */
    104, 105, 234, 235, 236, 237, 238, 239,     /* 232 - 239 */
    112, 113, 114, 115, 116, 117, 118, 119,     /* 240 - 247 */
    120,  62, 250, 251, 252, 253, 254, 255,     /* 248 - 255 */
  };

/* This is almost identical to GNU dd's "ibm" table.  In dd's "ibm"
   table, both 91 and 213 recode to 173, both 93 and 229 recode to 189,
   and no character recodes to 74 or 106.  I suspect two errors in there.
   So, I arbitrarily choose to recode 213 by 74 and 229 by 106.  */

static unsigned char const ascii_to_ebcdic_ibm[256] =
  {
      0,   1,   2,   3,  55,  45,  46,  47,     /*   0 -   7 */
     22,   5,  37,  11,  12,  13,  14,  15,     /*   8 -  15 */
     16,  17,  18,  19,  60,  61,  50,  38,     /*  16 -  23 */
     24,  25,  63,  39,  28,  29,  30,  31,     /*  24 -  31 */
     64,  90, 127, 123,  91, 108,  80, 125,     /*  32 -  39 */
     77,  93,  92,  78, 107,  96,  75,  97,     /*  40 -  47 */
    240, 241, 242, 243, 244, 245, 246, 247,     /*  48 -  55 */
    248, 249, 122,  94,  76, 126, 110, 111,     /*  56 -  63 */
    124, 193, 194, 195, 196, 197, 198, 199,     /*  64 -  71 */
    200, 201, 209, 210, 211, 212, 213, 214,     /*  72 -  79 */
    215, 216, 217, 226, 227, 228, 229, 230,     /*  80 -  87 */
    231, 232, 233, 173, 224, 189,  95, 109,     /*  88 -  95 */
    121, 129, 130, 131, 132, 133, 134, 135,     /*  96 - 103 */
    136, 137, 145, 146, 147, 148, 149, 150,     /* 104 - 111 */
    151, 152, 153, 162, 163, 164, 165, 166,     /* 112 - 119 */
    167, 168, 169, 192,  79, 208, 161,   7,     /* 120 - 127 */
     32,  33,  34,  35,  36,  21,   6,  23,     /* 128 - 135 */
     40,  41,  42,  43,  44,   9,  10,  27,     /* 136 - 143 */
     48,  49,  26,  51,  52,  53,  54,   8,     /* 144 - 151 */
     56,  57,  58,  59,   4,  20,  62, 225,     /* 152 - 159 */
     65,  66,  67,  68,  69,  70,  71,  72,     /* 160 - 167 */
     73,  81,  82,  83,  84,  85,  86,  87,     /* 168 - 175 */
     88,  89,  98,  99, 100, 101, 102, 103,     /* 176 - 183 */
    104, 105, 112, 113, 114, 115, 116, 117,     /* 184 - 191 */
    118, 119, 120, 128, 138, 139, 140, 141,     /* 192 - 199 */
    142, 143, 144, 154, 155, 156, 157, 158,     /* 200 - 207 */
    159, 160, 170, 171, 172,  74, 174, 175,     /* 208 - 215 */
    176, 177, 178, 179, 180, 181, 182, 183,     /* 216 - 223 */
    184, 185, 186, 187, 188, 106, 190, 191,     /* 224 - 231 */
    202, 203, 204, 205, 206, 207, 218, 219,     /* 232 - 239 */
    220, 221, 222, 223, 234, 235, 236, 237,     /* 240 - 247 */
    238, 239, 250, 251, 252, 253, 254, 255,     /* 248 - 255 */
  };

static void
init_ascii_ebcdic (STEP *step)
{
  step->one_to_one = ascii_to_ebcdic;
}

static void
init_ebcdic_ascii (STEP *step)
{
  step->one_to_one = invert_table (ascii_to_ebcdic);
}

static void
init_ascii_ebcdic_ccc (STEP *step)
{
  step->one_to_one = ascii_to_ebcdic_ccc;
}

static void
init_ebcdic_ccc_ascii (STEP *step)
{
  step->one_to_one = invert_table (ascii_to_ebcdic_ccc);
}

static void
init_ascii_ebcdic_ibm (STEP *step)
{
  step->one_to_one = ascii_to_ebcdic_ibm;
}

static void
init_ebcdic_ibm_ascii (STEP *step)
{
  step->one_to_one = invert_table (ascii_to_ebcdic_ibm);
}

void
module_ebcdic (void)
{
  declare_step (ASCII, "EBCDIC", REVERSIBLE, init_ascii_ebcdic,
		file_one_to_one);
  declare_step ("EBCDIC", ASCII, REVERSIBLE, init_ebcdic_ascii,
		file_one_to_one);
  declare_step (ASCII, "EBCDIC-CCC", REVERSIBLE, init_ascii_ebcdic_ccc,
		file_one_to_one);
  declare_step ("EBCDIC-CCC", ASCII, REVERSIBLE, init_ebcdic_ccc_ascii,
		file_one_to_one);
  declare_step (ASCII, "EBCDIC-IBM", REVERSIBLE, init_ascii_ebcdic_ibm,
		file_one_to_one);
  declare_step ("EBCDIC-IBM", ASCII, REVERSIBLE, init_ebcdic_ibm_ascii,
		file_one_to_one);
}
